xtask/src/commands/release/apply_plan.rs

```diff
@@ -0,0 +1,200 @@
+    let branch_name = "release-branch"; // Placeholder for the actual branch name logic
+    let branch_name = format!(
+        "{}-{}",
+        branch_name,
+        jiff::Timestamp::now().strftime("%Y-%m-%d")
+    );
+
+    // Switch to the new branch
+    if dry_run {
+        println!("Dry run: would create a new branch: {}", branch_name);
+    } else {
+        // Create a new branch
+        let status = Command::new("git")
+            .arg("switch")
+            .arg("-c")
+            .arg(&branch_name)
+            .status()
+            .context("Failed to create new branch")?;
+
+        if !status.success() {
+            bail!("Failed to create new branch: {}", branch_name);
+        }
+    }
+
+    // Commit the changes
+    if dry_run {
+        println!("Dry run: would commit changes to branch: {}", branch_name);
+    } else {
+        Command::new("git")
+            .arg("commit")
+            .arg("-am")
+            .arg("Finalize crates for release")
+            .status()
+            .context("Failed to commit changes")?;
+    }
+
+    // Push the branch
+    let url = if dry_run {
+        println!("Dry run: would push branch: {}", branch_name);
+        String::from("<dry_run_url>")
+    } else {
+        // git push origin <branch_name>
+        let message = Command::new("git")
+            .arg("push")
+            .arg("origin")
+            .arg(&branch_name)
+            .output()
+            .context("Failed to push branch")?;
+
+        if !message.status.success() {
+            bail!(
+                "Failed to push branch: {}",
+                String::from_utf8_lossy(&message.stderr)
+            );
+        }
+
+        // Extract the URL from the output
+        String::from_utf8_lossy(&message.stderr) // git outputs to stderr
+            .lines()
+            .map(|s| s.trim_start_matches("remote:"))
+            .map(|s| s.trim())
+            .find(|&s| s.starts_with("https://"))
+            .unwrap_or("")
+            .to_string()
+    };
+
+    // Open a pull request
+
+    let packages_to_release = release_plan
+        .packages
+        .iter()
+        .map(|step| format!("- {}: {}", step.package, step.new_version))
+        .collect::<Vec<_>>()
+        .join("\n");
+
+    let body = format!(
+        r#"This pull request prepares the following packages for release:
+
+{packages_to_release}
+
+The release plan used for this release:
+
+```json
+{release_plan_str}
+```
+        "#
+    );
+
+    let pr_url_base = if url.starts_with("https://github.com/esp-rs/") {
+        format!("https://github.com/esp-rs/esp-hal/compare/{branch_name}")
+    } else {
+        let Some(user) = url.split('/').nth(3) else {
+            bail!("Failed to extract user from URL: {}", url);
+        };
+        format!("https://github.com/esp-rs/esp-hal/compare/main...{user}:esp-hal:{branch_name}")
+    };
+
+    let open_pr_url = format!(
+        "{}?quick_pull=1&title=Prepare+release&body={}",
```

Comment:
Can we add labels via this?

What I'd really like to do is add a label that triggers a full docs run on the preview server, this would stop us from merging the PR with the docs broken (something we keep doing, and something we'll likely to forget to check again).

Something like a `release-pr` label, which we can attach some special logic to further test things like docs before we commit to the release.

Reply:
Yes! I'll also drop a comment in the code that the options are documented at https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/using-query-parameters-to-create-a-pull-request